File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation


def recv_all(socket: socket.socket, size):
    data = b""
    while len(data) < size:
        chunk = socket.recv(size - len(data))
        if not chunk:
            # aca habria algun error
            return None
        data += chunk
    return data

def send_all(socket: socket.socket, bytes):
    total = 0
    while total < len(bytes):
        n = socket.send(bytes[total:])
        if n == 0:
            # aca habria algun error
            return None
        total += n
    return total

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	for len(bytes) > 0 {
		n, err := socket.Write(bytes)
		if err != nil {
			return err
		}
		bytes = bytes[n:]
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	r := 0
	for r < size {
		n, err := socket.Read(buff[r:])
		if err != nil {
			return nil, err
		}
		r += n
	}
	return buff[:n], nil
}
